src/slice-assign-array.c

```diff
@@ -1,271 +1,425 @@
+                                                                            \
+  struct strides_info info = new_strides_info(proxy, index);                \
+  struct strides_info* p_info = &info;                                      \
+  PROTECT_STRIDES_INFO(p_info, &n_protect);                                 \
+                                                                            \
+  SEXP out = PROTECT_N(vec_clone_referenced(proxy, ownership), &n_protect); \
+  CTYPE* p_out = DEREF(out);                                                \
+                                                                            \
+  const R_len_t start = p_info->p_index[0];                                 \
+  const R_len_t step = p_info->p_index[2];                                  \
+                                                                            \
+  const CTYPE* p_value = CONST_DEREF(value);                                \
+  R_len_t k = 0;                                                            \
+                                                                            \
+  for (R_len_t i = 0; i < p_info->shape_elem_n; ++i) {                      \
+    R_len_t loc = vec_strided_loc(                                          \
+      p_info->p_shape_index,                                                \
+      p_info->p_strides,                                                    \
+      p_info->shape_n                                                       \
+    );                                                                      \
+                                                                            \
+    loc += start;                                                           \
+                                                                            \
+    for (R_len_t j = 0; j < p_info->index_n; ++j) {                         \
+      p_out[loc] = p_value[k];                                              \
+      loc += step;                                                          \
+      k += VALUE_POST_INDEX_INCREMENT;                                      \
+    }                                                                       \
+                                                                            \
+    vec_shape_index_increment(p_info);                                      \
+    k += VALUE_POST_SHAPE_INCREMENT;                                        \
+  }                                                                         \
+                                                                            \
+  UNPROTECT(n_protect);                                                     \
   return out
 
-#define ASSIGN_SHAPED_COMPACT(                                         \
+// -----------------------------------------------------------------------------
+
+#define ASSIGN_SHAPED_CONDITION_INDEX(                                 \
```

Comment:
You can just trust me that this is right if you want, I've added tests, but it's actually much simpler than the location case, we don't even need the strides information.